### PR TITLE
Minify production build image

### PR DESCRIPTION
**Pull request by tpinney**
### Context

The current build assets are not compressed and coupled with the way the docker image is built results in a final build image of 586MB at present and it will only get larger. The aim of this task is to ensure the production assets in the production build are compressed and any unnecessary build components are not built into the production image.

The resultant image size from this work is 164MB, approximately a 70% reduction :-)

### Changes proposed in this pull request

- The dockerfile has been change to a multi-stage dockerfile comprising of four separate stages.
  - Stage 1 - Contains the common build components to both development and production
  - Stage 2 - For development builds only and builds on stage one by copying the application code in and sets the entrypoint
  - Stage 3 - For production builds, this is an intermediate stage that builds on stage one copying the application into the image, compiling the assets and then removing any unnecessary files from the image.
  - Stage 4 - The final part of the production image build starts with a fresh image that installs the bare minimum dependencies and then copied in the application code and compiled assets from stage three.
- The docker compose override files have had build args specific to their environments added and a target for the desired build stage
- Rakefile has had a minor tweak to ensure that the entity diagram in the documentation is only run under development
- Makefile has had the test step modified to install the test dependencies in the container during test only so that it doesn't feature unnecessarily in the production image hosted in the Azure application service container.

### Guidance to review

This change could benefit from one or more of the developers to run a dev build from this branch on their local machine to ensure everything works as expected. The build completes successfully when RAILS_ENV=development but I don't know if I've inadvertently broken anything.

### Link to Trello card

[1013 - Minify assets in production](https://trello.com/c/WV0XHND1/1013-minify-assets-in-production)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,24 +1,102 @@
...
+RUN bundle install && \
+    yarn install
+
+
+#### Dev build builds on common build env image ####
+
+FROM common-build-env AS dev-build
+ARG APP_HOME=/app
+ARG railsEnv="development"
+ENV RAILS_ENV=${railsEnv}
```
Comment (duncanjbrown) [inline Dockerfile]: I don't understand what the ARG is for here: the stage is called `dev-build` and it's explicitly made the target in `docker-compose.override.yml`. Why not hardcode `ENV RAILS_ENV="development"`? (Also applies to the `railsEnv` in `prod-minify`)

File: Dockerfile [hunk 2]
```diff
@@ -1,24 +1,102 @@
...
 
 WORKDIR $APP_HOME
-COPY Gemfile Gemfile.lock package.json yarn.lock $APP_HOME/
 
-RUN gem install bundler:2.0.2 && \
-    yarn install --check-files && \
-    bundle install
+RUN rm -rf $BUNDLE_PATH/cache/*.gem && \
+    find $BUNDLE_PATH/gems/ -name "*.c" -delete && \
+    find $BUNDLE_PATH/gems/ -name "*.o" -delete
```
Comment (duncanjbrown) [inline Dockerfile]: If we're purging C source, how about `*.h` too?

File: Dockerfile [hunk 3]
```diff
@@ -1,24 +1,102 @@
...
 
-COPY . $APP_HOME/
+COPY . .
```
Comment (duncanjbrown) [inline Dockerfile]: Isn't this already inside the image if we're taking `common-build-env`?

File: Dockerfile [hunk 4]
```diff
@@ -1,24 +1,102 @@
...
 
 RUN bundle exec rake assets:precompile
 
+RUN rm -rf tmp/cache app/assets lib/assets vendor/assets node_modules
```
Comment (duncanjbrown) [inline Dockerfile]: 👌

File: Rakefile [hunk 5]
```diff
@@ -17,5 +17,5 @@ end
 task default: %i[lint_erb lint_ruby spec generate_state_diagram brakeman]
 
 Rake::Task['db:migrate'].enhance do
-  sh 'bundle exec erd'
+  sh 'bundle exec erd' if Rails.env.development?
```
Comment (duncanjbrown) [inline Rakefile]: 👍

File: Makefile [hunk 6]
```diff
@@ -34,4 +34,6 @@ ci.lint-erb: ## Run the ERB linter
 
 .PHONY: ci.test
 ci.test: ## Run the tests with results formatted for CI
-	docker-compose run --rm web /bin/sh -c 'bundle exec --verbose rspec --format RspecJunitFormatter' > rspec-results.xml
+	docker-compose run --rm web /bin/sh -c 'apk add nodejs yarn && \
+		bundle exec rails assets:precompile && \
```
Comment (duncanjbrown) [inline Makefile]: Interesting. it shouldn't be necessary to precompile in order to run the specs.

File: Dockerfile [hunk 7]
```diff
@@ -1,24 +1,102 @@
...
+
+FROM ruby:2.6.3-alpine AS prod-build
+ARG railsEnv="production"
+ARG bundleWithout=""
+ARG APP_HOME=/app
+ARG PACKAGES="tzdata postgresql-client"
+
+ENV RAILS_ENV=${railsEnv}
+ENV BUNDLE_WITHOUT=${bundleWithout}
+ENV BUNDLE_PATH="/gems"
```
Comment (duncanjbrown) [inline Dockerfile]: Again, as this depends on prior images, can we get away with declaring it once at the top?

File: Dockerfile [hunk 8]
```diff
@@ -1,24 +1,102 @@
...
+COPY Gemfile Gemfile.lock package.json yarn.lock ./
+
+RUN bundle install && \
+    yarn install
+
+
+#### Dev build builds on common build env image ####
+
+FROM common-build-env AS dev-build
+ARG APP_HOME=/app
```
[the remark this answers is not in the document]
Comment (tpinney) [reply Dockerfile]: When I switched to multistage Docker images it didn't seem to accept re-use of the APP_HOME variable after a new FROM statement, so I had to repeat it unfortunately.

File: Dockerfile (hunk 1 above)
Comment (tpinney) [reply Dockerfile]: The ARG is the only way you can pass in variables into a docker image at build time. I can happily hardcode it at each stage if you want me to, but I felt this gave us the ability to override the default value for RAILS_ENV should the need arise in the future.
Let me know you preference and I'll make the change if required.

File: Dockerfile (hunk 2 above)
Comment (tpinney) [reply Dockerfile]: That's a good question. I'm not a ruby/rails expert and all the guides I found on the web regarding removal of intermediate files all only mentions .c and .o files so I never even considered the header files. I'll give it a try without the header files and see what effect if any its has.

File: Dockerfile (hunk 3 above)
Comment (tpinney) [reply Dockerfile]: The <common-build-env> only has the Gemfile, packages.json and yarn.lock file, so the app code needs to be copied in at this stage.

File: Dockerfile (hunk 7 above)
Comment (tpinney) [reply Dockerfile]: Because of the repeated use of FROM we can't :(

File: Makefile (hunk 6 above)
Comment (tpinney) [reply Makefile]: I just double checked on this one and this is the error I get when I remove the assets:precompile option.
```
--- Caused by: ---
Webpacker::Manifest::MissingEntryError:
  Webpacker can&apos;t find media/images/favicon.ico in /app/public/packs-test/mm
anifest.json. Possible causes:
  1. You want to set webpacker.yml value of compile to true for your environment
     unless you are using the `webpack -w` or the webpack-dev-server.
  2. webpack has not yet re-run to reflect updates.
  3. You have misconfigured Webpacker&apos;s config/webpacker.yml file.
  4. Your webpack configuration is not creating a manifest.
```
I assumed (2) was the cause, hence the presence of the precompile option. Should I have done something else to resolve this?

File: Dockerfile (hunk 2 above)
Comment (tpinney) [reply Dockerfile]: Removal of the C header files appears not to have had any detrimental effects, so I've updated the branch to include this change too.

File: Dockerfile (hunk 1 above)
Comment (duncanjbrown) [reply Dockerfile]: I think I prefer it hardcoded as it's more to process when reading the file. Could we add it when we first need the feature?

File: Dockerfile (hunk 2 above)
Comment (duncanjbrown) [reply Dockerfile]: :+1: great

File: Dockerfile (hunk 3 above)
Comment (duncanjbrown) [reply Dockerfile]: My mistake, I misread `dev-build` as part of `common-build-env`

File: Dockerfile (hunk 7 above)
Comment (duncanjbrown) [reply Dockerfile]: gah

File: Makefile (hunk 6 above)
Comment (duncanjbrown) [reply Makefile]: It's clearly using the test env as the path points to `packs-test`. As I understand the out of the box behaviour in test env it dynamically builds the webpacker assets when they're first released. I have no idea why this isn't working in quite that way, though. Don't see much point in blocking on this

File: Dockerfile (hunk 1 above)
Comment (tpinney) [reply Dockerfile]: @duncanjbrown I have committed this change which is currently building.

**Pull request merged**